pkg/analytics/categories.go

```diff
@@ -0,0 +1,127 @@
+package analytics
+
+import (
+	"slices"
+	"strings"
+)
+
+type CliArgCategorizer struct {
+	commands []string // parsed commands
+	flags    []string // parsed flags
+	result   []string // final categories passed to output
+
+	flagsWhitelist  []string
+	knownCommands   []string
+	knownProducts   []string // {"sbom", "code", "iac", "container"}
+	productFallback string   // {"oss"}
+}
+
+// NewCategoryParser creates a new instance of CliArgCategorizer, which is used to
+// parse and categorize command-line arguments based on known commands, whitelisted flags,
+// and recognized products.
+//
+// Parameters:
+//   - knownCommands: A slice of strings representing valid main commands and subcommands.
+//   - flagsWhitelist: A slice of strings representing whitelisted flags (without leading dashes).
+//   - knownProducts: A slice of strings representing the names of recognized products.
+//
+// Returns:
+//   - A pointer to a new CliArgCategorizer instance, ready to be used for parsing arguments.
+//
+// Usage:
+// categorizer := NewCategoryParser(knownCommands, whitelistedFlags, knownProducts)
+// categoryVector = categorizer.CategorizeCliArgs(os.Args)
+func NewCategoryParser(knownCommands []string, flagsWhitelist []string, knownProducts []string) *CliArgCategorizer {
+	return &CliArgCategorizer{
+		commands: []string{},
+		flags:    []string{},
+		result:   []string{},
+
+		flagsWhitelist:  flagsWhitelist,
+		knownCommands:   knownCommands,
+		knownProducts:   knownProducts,
+		productFallback: "oss",
+	}
+}
+
+// CategorizeCliArgs parses and categorizes command-line arguments into a structured slice.
+// The categorization follows a specific order: [product, command(s), flags].
+//
+// This function processes the arguments starting from index 1 (assuming the first argument is the CLI name).
+// It identifies the main command, any subcommands, and whitelisted flags.  It then constructs a category vector
+// where:
+//   - The first element is the product category (e.g., "oss", "code", "iac"), defaulting to "oss" if not explicitly specified.
+//   - The second element (or more) is the main command, potentially followed by any valid subcommands.
+//   - The remaining elements are whitelisted flags (without the leading dashes).
+//
+// Parameters:
+//   - args: A slice of strings representing the command-line arguments.
+//
+// Returns:
+//   - A slice of strings representing the categorized arguments.
+func (c *CliArgCategorizer) CategorizeCliArgs(args []string) []string {
```

Comment:
nitpick: Naming is hard but since this is framed to be a parser, why not call this `parse()`